cpp/src/arrow/array/array_nested.cc

```diff
@@ -469,6 +469,69 @@ inline void SetListData(VarLengthListLikeArray<TYPE>* self,
   self->values_ = MakeArray(self->data_->child_data[0]);
 }
 
+Result<std::shared_ptr<Array>> FlattenLogicalListRecursively(const Array& array,
+                                                             MemoryPool* memory_pool) {
+  Type::type kind = array.type_id();
+  std::shared_ptr<Array> in_array = array.Slice(0, array.length());
+  while (is_list_like(kind) || is_list_view(kind)) {
+    const bool has_nulls = array.null_count() > 0;
+    std::shared_ptr<Array> out;
```

Comment:
You don't need this temporary variable. You can use and assign on `array` on the same statement. Less local variables == less destructors to call at the end.